### DISCUSSION TITLE
How to force a rebuild of mpfs icicle dev kit's devicetrees

### DISCUSSION THREAD

**Discussion by es-ggg** [polarfire-soc/polarfire-soc-documentation#388] [How to] [answered]
Hello,

I'm not familiar with how to force a device tree rebuild after it's been modified. For more information on exactly what I'm trying to do, please read on.

I want to attempt modifying the device trees of the icicle dev kit, because I'll eventually need to add more uio devices to be able to communicate to multiple modules from user space, so I decided to take the approach of deleting a node from the device tree to see if when I attempt rebuilding the project by using the command **MACHINE=icicle-kit-es bitbake mpfs-dev-cli**, a new image with the modification is generated. The first thing I observed was that a new image was not generated, and the .dtb files I located were not rebuilt either. I wrote the image to the SD card anyway and the node I attempted deleting was still in present on the unit's  /dev/ 

The exact node I attempted deleting was the udmabuf0 node located here: ~/yocto-dev/build/tmp-glibc/work-shared/icicle-kit-es/kernel-source/arch/riscv/boot/dts/microchip/mpfs-icicle-kit.dts

After deleting the node, I rebuilt with the command **MACHINE=icicle-kit-es bitbake mpfs-dev-cli** and got no new image or dtb.

Could I please be directed on how to force a device tree binary/blob rebuild that can then be used in generating a new image that can be written to the SD card?

Thanks in advance!

**Comment by rockoguido** [comment, marked as answer]
Hi..  if you are trying to modify the Linux device tree for a Yocto based build there are a number of steps you need to go through to create a patch and then have the patch applied, you cannot modify the device tree directly and do a rebuild, that change, as you noted, will not stick.
1. Can we assume you have created a Linux based patch file in the past?  I'll just list high level steps here...
2. go to the location of the device tree,   see mpfs-icicle-kit-dts in ...arch/riscv/boot/dts/microchip.  enter a git status command
3. results should indicate the branch you are using and if any changes are in place
4.  edit the dts file
5. enter git diff command followed by git status command
6.  should reflect change you just made
7. now enter a git add .               command  you can follow with a git status command just to check results
8. git commit -s -m "<patch name>"
9. now git format-patch -1
10. above command creates the patch file that is needed to modify the device tree
11. this file needs to be moved to your yocto root directory and then to meta-polarfire-soc-yocto-bsp/recipes-kernel/linux/files/icicle-kit-es$
12. in meta-polarfire-soc-yocto-bsp/recipes-kernel/linux/     directory you need to update mpfs-linux.bb to reflect the patch
13. an example lies below..
14. 
15. SRC_URI:append:icicle-kit-es = " file://mpfs_cmdline.cfg \
    file://rpi_sense_hat.cfg \
    file://qspi_flash.cfg \
    file://0001-Mem-Reduction-to-1Gig.patch \
    file://0001-usb-perihperal.patch \

then you are free to build, I would do a bitbake -c cleansstate just to be sure you rebuild with your patch.
also would recommend building   sequence could look like.. 

 MACHINE=icicle-kit-es bitbake core-image-minimal-dev -c cleansstate
MACHINE=icicle-kit-es bitbake core-image-minimal-dev

So, there's a few steps, take your time and verify them all
Regards,
RG

**Reply by xiaogu-arch** [reply]
Hello, I have a question. So, if one wants to revert the device tree to its original source code state, how should to proceed?

**Reply by xiaogu-arch** [reply]
@rockoguido hello,RG！
I would like to know how to modify the device tree of the "MACHINE=icicle-kit-es-nandI" image. Could you please tell me what to do? I couldn't find the "icicle-kit-es-nand" folder under the path of "meta-polarfire-soc-yocto-bsp/recipes-kernel/linux/files".

**Comment by es-ggg** [comment]
Thank you so much for you quick response RG!! It works exactly how you described it. I successfully made the modifications I wanted.
